### PR TITLE
Add `F.roi_max_pooling_2d`

### PR DESCRIPTION
**Pull request by knorth55**
- add new interface function `roi_max_pooling_2d`
- add doc
- add test
roi_max_pooling_2d

this is related to #5285, #5198.

### CODE REVIEW

**Comment by knorth55** [comment]
@Hakuyume  kindly ping. if possible, can you review this?

**Comment by knorth55** [comment]
@Hakuyume can you review this PR?

File: chainer/functions/pooling/roi_max_pooling_2d.py [hunk 1]
```diff
@@ -0,0 +1,373 @@
...
+
+            # iterate all the w, h (from feature map) that fall into this ROIs
+            for w in six.moves.range(xmin, xmax + 1):
+                for h in six.moves.range(ymin, ymax + 1):
+                    phstart = int(numpy.floor(float(h - ymin) / strideh))
+                    phend = int(numpy.ceil(float(h - ymin + 1) / strideh))
+                    pwstart = int(numpy.floor(float(w - xmin) / stridew))
+                    pwend = int(numpy.ceil(float(w - xmin + 1) / stridew))
+
+                    phstart = min(max(phstart, 0), self.outh)
```
Comment (Hakuyume) [inline chainer/functions/pooling/roi_max_pooling_2d.py]: Do we need to compare `phstart` with 0?

Comment (knorth55) [reply chainer/functions/pooling/roi_max_pooling_2d.py]: I just followed https://github.com/chainer/chainer/blob/master/chainer/functions/pooling/roi_pooling_2d.py.
IMO, we do not need to compare, but there might be some exception.
Also, I looked into Caffe2, and they have completely different but simple kernel.
Should we move to this simple one?
https://github.com/pytorch/pytorch/blob/master/caffe2/operators/roi_pool_op.cu#L101-L119

Comment (knorth55) [reply chainer/functions/pooling/roi_max_pooling_2d.py]: I updated to follow caffe2 kernel.

**Comment by knorth55** [comment]
@Hakuyume can you review this?

**Comment by Hakuyume** [comment]
jenkins, test this please

**Comment by kmaehashi** [comment]
Jenkins, test this please.

File: chainer/functions/pooling/roi_max_pooling_2d.py [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+
+from chainer.backends import cuda
+from chainer import function
+from chainer.utils import type_check
+
+from chainer.functions.pooling.roi_pooling_2d import _roi_pooling_slice
+
+
+def _pair(x):
+    if isinstance(x, collections.Iterable):
```
Comment (kmaehashi) [inline chainer/functions/pooling/roi_max_pooling_2d.py]: For compatibility with Python 3.7, please use `chainer.utils.collections_abc` instead of `collections`.

**Comment by knorth55** [comment]
Any updates?

**Comment by kmaehashi** [comment]
@Hakuyume Could you confirm the changes since your last review?

File: chainer/functions/pooling/roi_max_pooling_2d.py [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+        outh, outw = _pair(outsize)
+        if not (isinstance(outh, int) and outh > 0):
+            raise TypeError(
+                'outsize[0] must be positive integer: {}, {}'
+                .format(type(outh), outh))
+        if not (isinstance(outw, int) and outw > 0):
+            raise TypeError(
+                'outsize[1] must be positive integer: {}, {}'
+                .format(type(outw), outw))
+        if isinstance(spatial_scale, int):
```
Comment (Hakuyume) [inline chainer/functions/pooling/roi_max_pooling_2d.py]: This check accepts negative int (e.g. `-1`). In addition, it cannot take number-like types (e.g. 'numpy.float32').
How about this?
```python
try:
    spatial_scale = float(spatial_scale)
except ValueError:
    # raise an error for casting
if spatial_scale <= 0:
    # raise an error for non-positive value
```

Comment (knorth55) [reply chainer/functions/pooling/roi_max_pooling_2d.py]: > In addition, it cannot take number-like types (e.g. 'numpy.float32').

We discussed this problem locally and decided not to support `numpy.scalar` in this PR.

Comment (knorth55) [reply chainer/functions/pooling/roi_max_pooling_2d.py]: @Hakuyume BTW, i found that `numpy.float64` works.

```python
Python 2.7.12 (default, Dec  4 2017, 14:50:18) 
Type "copyright", "credits" or "license" for more information.

IPython 5.5.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import numpy as np

In [2]: a = np.float64(1.6)

In [3]: isinstance(a, float)
Out[3]: True

In [4]: b = np.float32(1.6)

In [5]: isinstance(b, float)
Out[5]: False

```

Comment (Hakuyume) [reply chainer/functions/pooling/roi_max_pooling_2d.py]: :tada:

**Comment by Hakuyume** [comment]
Jenkins, test this please.

File: chainer/functions/pooling/roi_max_pooling_2d.py (hunk 3 above)
Comment (knorth55) [reply chainer/functions/pooling/roi_max_pooling_2d.py]: `np.issubtype` seems good for checking scalar value.
https://docs.scipy.org/doc/numpy-1.15.1/reference/generated/numpy.issubdtype.html

File: tests/chainer_tests/functions_tests/pooling_tests/test_roi_max_pooling_2d.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+import chainer
+from chainer.backends import cuda
+from chainer import functions
+from chainer import gradient_check
+from chainer import testing
+from chainer.testing import attr
+
+
+def _pair(x):
+    if isinstance(x, collections.Iterable):
```
Comment (Hakuyume) [inline tests/chainer_tests/functions_tests/pooling_tests/test_roi_max_pooling_2d.py]: This line should be fixed.

File: chainer/functions/pooling/roi_max_pooling_2d.py (hunk 3 above)
Comment (Hakuyume) [reply chainer/functions/pooling/roi_max_pooling_2d.py]: It raises a warning for `float`.
```python
>>> np.issubdtype(type(1.0), float)
/usr/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
```

File: tests/chainer_tests/functions_tests/pooling_tests/test_roi_max_pooling_2d.py (hunk 4 above)
Comment (knorth55) [reply tests/chainer_tests/functions_tests/pooling_tests/test_roi_max_pooling_2d.py]: Thx! Updated.

**Comment by Hakuyume** [comment]
Jenkins, test this please.

**Comment by Hakuyume** [comment]
Jenkins failed on the test of `BatchNormalization`. It is not related to this PR.
@kmaehashi How should I treat this problem?

File: chainer/functions/pooling/roi_max_pooling_2d.py (hunk 3 above)
Comment (knorth55) [reply chainer/functions/pooling/roi_max_pooling_2d.py]: how about this?
```python
np.issubdtype(type(1.0), np.floating)
# no warnings
np.issubdtype(type(1), np.integer)
# no warnings
```

**Comment by knorth55** [comment]
Should i merge master branch and wait for jenkins test again?

**Comment by kmaehashi** [comment]
Thanks for resolving conflicts.

> Should i merge master branch and wait for jenkins test again?

Jenkins applies this PR against master branch and tests it, so you don't have to do that unless there is a conflict.

Jenkins, test this please.
